ngx_http_sxg_filter_module.c

```diff
@@ -303,38 +326,114 @@ static ngx_str_t extract_angled_url(char* str, size_t len) {
   return (ngx_str_t){.data = NULL, .len = 0};
 }
 
-static void extract_preaload_url_list(ngx_str_t* link, ngx_array_t* const dst,
-                                      ngx_http_request_t* r) {
+bool concat_str(ngx_str_t* target, const char* buff, size_t len,
+                ngx_pool_t* pool) {
+  u_char* new_buffer = ngx_palloc(pool, target->len + len);
+  if (new_buffer == NULL) {
+    return false;
+  }
+  ngx_memcpy(new_buffer, target->data, target->len);
+  ngx_memcpy(new_buffer + target->len, buff, len);
+  ngx_pfree(pool, target->data);
+  target->data = new_buffer;
+  target->len += len;
+  return true;
+}
+
+// TODO(kumagi): Complex logic should be migrated to ngx_sxg_utils.c.
+static void extract_preload_url_list(ngx_str_t* link, ngx_array_t* const dst,
+                                     ngx_str_t* non_preload_headers,
+                                     ngx_http_request_t* r) {
   char* str = (char*)link->data;
-  char* const end = str + link->len;
+  char* end = str + link->len;
   while (str < end) {
     const size_t tail = get_term_length(str, end - str, ',', "<>");
     ngx_str_t url = extract_angled_url(str, tail);
     char* param = str;
+    bool found = false;
+    ngx_subresource_t* new_preload = NULL;
+    ngx_str_t as = ngx_null_string;
     while (param < str + tail) {
       const size_t rest = str + tail - param;
       size_t param_tail = get_term_length(param, rest, ';', "<>");
       if (param_is_preload(param, param_tail)) {
-        ngx_str_t* const new_url = ngx_array_push(dst);
-        *new_url = url;
-        break;
+        new_preload = ngx_array_push(dst);
+        new_preload->url.data = ngx_palloc(r->pool, url.len);
+        new_preload->url.len = url.len;
+        ngx_memcpy(new_preload->url.data, url.data, url.len);
+        while (new_preload->url.data[new_preload->url.len - 1] == ' ') {
+          new_preload->url.len--;
+        }
+        found = true;
+
+        if (as.data != NULL) {
+          new_preload->as = as;
+        }
+      }
+      if (param_is_as(param, param_tail, (const char**)&as.data, &as.len)) {
+        if (found) {
+          new_preload->as = as;
+          ngx_log_error(NGX_LOG_DEBUG, r->connection->log, 0,
+                        "found as statement in link preload header: %V",
+                        &new_preload->as);
+        }
       }
       param += param_tail + 1;
     }
+    if (!found) {
+      if (non_preload_headers->len > 0) {
+        concat_str(non_preload_headers, ",", 1, r->pool);
+      }
+      concat_str(non_preload_headers, str, tail, r->pool);
```

Comment:
How many headers do you expect to have?  This has O(n**2) behavior due to the allocations and memcpy.  Maybe better to accumulate an `ngx_array_t` and create a single string after the loop?  Or loop twice, once getting the length and against copying the bytes, like `vsnprintf` does.

Reply:
OK, I modified the function to work in 2 modes, estimation only and actual write. First run this function in estimation mode and get the expected memory length, and allocate and do actual write.